src/dbnode/integration/index_block_orphaned_entry_test.go

```diff
@@ -0,0 +1,207 @@
+	for i := 0; i < numTestSeries; i++ {
+		fooID := ident.StringID(fmt.Sprintf("foo.%v", i))
+		ids = append(ids, fooID)
+
+		writeConcurrentMetrics(t, setup, session, fooID)
+	}
+
+	// Write metrics for a different series to push current foreground segment
+	// to the background. After this, all documents for foo.X exist in background segments
+	barID := ident.StringID("bar")
+	writeConcurrentMetrics(t, setup, session, barID)
+
+	// Fast-forward to a block rotation
+	newBlock := xtime.Now().Truncate(blockSize).Add(blockSize)
+	newCurrentTime := newBlock.Add(30 * time.Minute) // Add extra to account for buffer past
+	setup.SetNowFn(newCurrentTime)
+
+	// Wait for flush
+	log.Info("waiting for block rotation to complete")
+	nsID := setup.Namespaces()[0].ID()
+	found := xclock.WaitUntil(func() bool {
+		filesets, err := fs.IndexFileSetsAt(setup.FilePathPrefix(), nsID, newBlock.Add(-blockSize))
+		require.NoError(t, err)
+		return len(filesets) == 1
+	}, 30*time.Second)
+	require.True(t, found)
+
+	// Do post-block rotation writes
+	for _, id := range ids {
+		writeMetric(t, session, nsID, id, newCurrentTime, 999.0)
+	}
+	writeMetric(t, session, nsID, barID, newCurrentTime, 999.0)
+
+	// Foreground segments should be in the background again which means updated index entry
+	// is now behind the orphaned entry so index reads should fail.
+	log.Info("waiting for metrics to be indexed")
+	var (
+		missing string
+		ok      bool
+	)
+	found = xclock.WaitUntil(func() bool {
+		for _, id := range ids {
+			ok, err = isIndexedCheckedWithTime(
+				t, session, nsID, id, genTags(id), newCurrentTime,
+			)
+			if !ok {
+				missing = id.String()
+				return false
+			}
+		}
+		return true
+	}, 30*time.Second)
+	assert.True(t, found, fmt.Sprintf("series %s never indexed\n", missing))
+	assert.NoError(t, err)
+}
+
+func writeConcurrentMetrics(
+	t *testing.T,
+	setup TestSetup,
+	session client.Session,
+	seriesID ident.ID,
+) {
+	var wg sync.WaitGroup
+	nowFn := setup.DB().Options().ClockOptions().NowFn()
+
+	workerPool := xsync.NewWorkerPool(concurrentWorkers)
+	workerPool.Init()
+
+	mdID := setup.Namespaces()[0].ID()
+	for i := 0; i < concurrentWorkers; i++ {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			for j := 0; j < writesPerWorker; j++ {
+				j := j
+				wg.Add(1)
+				workerPool.Go(func() {
+					defer wg.Done()
+					writeMetric(t, session, mdID, seriesID, xtime.ToUnixNano(nowFn()), float64(j))
+				})
+			}
+		}()
+	}
+
+	wg.Wait()
+}
+
+func genTags(seriesID ident.ID) ident.TagsIterator {
+	return ident.NewTagsIterator(ident.NewTags(ident.StringTag("tagName", seriesID.String())))
+}
+
+func writeMetric(
+	t *testing.T,
+	session client.Session,
+	nsID ident.ID,
+	seriesID ident.ID,
+	timestamp xtime.UnixNano,
+	value float64,
+) {
+	err := session.WriteTagged(nsID, seriesID, genTags(seriesID),
+		timestamp, value, xtime.Second, nil)
+	require.NoError(t, err)
+}
+
+func generateTestSetup(t *testing.T) TestSetup {
```

Comment:
Would be nice if we were able to namespace these method names. I guess if other tests want to use the same kind of thing they can in the future pass in their own overrides and we can move this to a shared place. e.g. `func generateTestSetup(t *testing.T, customOpts func(storage.Options) storage.Options) { ... }`.